<?xml version="1.0"?>
<launch>
    <arg name="device_model" value="hector_quadrotor" doc="This script only works for Hector Quadrotor"/>

    <!-- Configurations for starting the Gazebo simulation world -->
    <arg name="gui" default="false"/>
    <arg name="paused" default="true" doc="Default paused to wait for all nodes to initialize"/>
    <arg name="profile" default="default" doc="Physics profile in the world"/>
    <arg name="seed" default="1" doc="Random seed for Gazebo"/>
    <arg name="use_sim_time" default="true"/>

    <!-- Configurations for spawning a device -->
    <arg name="name" default="test_device0"/>
    <arg name= "x" default="0.0"/>
    <arg name= "y" default="0.0"/>
    <arg name= "z" default="0.3"/>

    <!-- Configuration for randomly generating waypoints -->
    <arg name="num_waypoints" default="10"/>
    <arg name="time_itvl_min" default="1.0"
         doc="Minimum time interval between assigning consecutive waypoints"/>
    <arg name="center" default="[0.0, 0.0, 2.5]"
         doc="Center for generating random points on a sphere"/>
    <arg name="dist_min" default="1.0"
         doc="Minimum distance between consecutive waypoints"/>
    <arg name="dist_max" default="5.0"
         doc="Maximum distance between consecutive waypoints"/>
    <arg name="x_min" default="0.0"/>
    <arg name="x_max" default="4.0"/>
    <arg name="y_min" default="0.0"/>
    <arg name="y_max" default="4.0"/>
    <arg name="z_min" default="0.5"/>
    <arg name="z_max" default="4.5"/>

    <!-- Configurations for recording a ROSBag-->
    <arg name="record_topics" default="ground_truth/state command/pose action/pose/goal"/>
    <arg name="duration" default="30m"/>
    <arg name="buffsize" default="256"/>
    <arg name="bag_dir" default="/tmp/"/>
    <arg name="bag_name_prefix" default="follow_waypoints"/>

    <!-- Create the world. -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
      <arg name="paused" value="$(arg paused)"/>
      <arg name="gui" value="$(arg gui)"/>
      <arg name="use_sim_time" value="$(arg use_sim_time)"/>
      <arg name="extra_gazebo_args" value="--profile $(arg profile) --seed $(arg seed)"/>
    </include>

    <!-- Spawn a device -->
    <include file="$(find cym_device)/launch/spawn_cym_device.launch">
        <arg name="name" value="$(arg name)"/>
        <arg name="device_model" value="$(arg device_model)"/>
        <arg name="x" value="$(arg x)"/>
        <arg name="y" value="$(arg y)"/>
        <arg name="z" value="$(arg z)"/>
    </include>

    <!-- Run Python testing scripts that assign random waypoints -->
    <node ns="$(arg name)" name="$(anon rand_gen_wps)" pkg="cym_device" type="test_hector_quad_random_waypoints.py"
          required="true">
        <param name="num_waypoints" type="int" value="$(arg num_waypoints)"/>
        <param name="time_itvl_min" type="double" value="$(arg time_itvl_min)"/>
        <rosparam param="center" subst_value="true">$(arg center)</rosparam>
        <param name="dist_min" type="double" value="$(arg dist_min)"/>
        <param name="dist_max" type="double" value="$(arg dist_max)"/>
        <param name="x_min" type="double" value="$(arg x_min)"/>
        <param name="x_max" type="double" value="$(arg x_max)"/>
        <param name="y_min" type="double" value="$(arg y_min)"/>
        <param name="y_max" type="double" value="$(arg y_max)"/>
        <param name="z_min" type="double" value="$(arg z_min)"/>
        <param name="z_max" type="double" value="$(arg z_max)"/>
    </node>

    <group ns="$(arg name)" if="$(eval arg('record_topics')!='')">
        <!-- ROS Bag record -->
        <node name="$(anon record)" pkg="rosbag" type="record" output="screen" required="true"
              args=" --buffsize $(arg buffsize) --duration $(arg duration) --tcpnodelay -o $(arg bag_dir)/$(arg bag_name_prefix) $(arg record_topics)"/>
    </group>
</launch>
